File: android/app/src/main/java/com/plantora/billing/ui/settings/staff/StaffViewModel.kt
package com.plantora.billing.ui.settings.staff

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.plantora.billing.data.SalespersonRepository
import com.plantora.billing.data.remote.friendlyError
import com.plantora.billing.domain.Salesperson
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.security.SecureRandom
import javax.inject.Inject

/** Result of a create/reset action — shows the one-time credentials. */
data class CredentialResult(val email: String, val password: String, val isReset: Boolean)

data class CreateForm(
    val email: String = "",
    val password: String = generatePassword(),
    val saving: Boolean = false,
    val error: String? = null,
) {
    val canSave: Boolean get() = email.contains("@") && password.length >= 8 && !saving
}

/** Reset a specific salesperson's password to a chosen (or generated) value. */
data class ResetForm(
    val sp: Salesperson,
    val password: String = generatePassword(),
    val saving: Boolean = false,
    val error: String? = null,
) {
    val canSave: Boolean get() = password.length >= 8 && !saving
}

data class StaffUiState(
    val loading: Boolean = true,
    val error: String? = null,
    val staff: List<Salesperson> = emptyList(),
    val createForm: CreateForm? = null,
    val resetForm: ResetForm? = null,
    val credentials: CredentialResult? = null,
    val message: String? = null,
)

@HiltViewModel
class StaffViewModel @Inject constructor(
    private val repo: SalespersonRepository,
) : ViewModel() {

    private val _ui = MutableStateFlow(StaffUiState())
    val ui: StateFlow<StaffUiState> = _ui.asStateFlow()

    init { load() }

    fun load() {
        _ui.update { it.copy(loading = true, error = null) }
        viewModelScope.launch {
            runCatching { repo.list() }
                .onSuccess { list -> _ui.update { it.copy(loading = false, staff = list) } }
                .onFailure { e -> _ui.update { it.copy(loading = false, error = friendlyError(e)) } }
        }
    }

    fun openCreate() = _ui.update { it.copy(createForm = CreateForm()) }
    fun closeCreate() = _ui.update { it.copy(createForm = null) }
    fun setEmail(v: String) = _ui.update { it.copy(createForm = it.createForm?.copy(email = v, error = null)) }
    fun setPassword(v: String) = _ui.update { it.copy(createForm = it.createForm?.copy(password = v, error = null)) }
    fun regeneratePassword() = _ui.update { it.copy(createForm = it.createForm?.copy(password = generatePassword(), error = null)) }
    fun dismissCredentials() = _ui.update { it.copy(credentials = null) }
    fun dismissMessage() = _ui.update { it.copy(message = null) }

    fun createStaff() {
        val form = _ui.value.createForm ?: return
        if (!form.canSave) return
        _ui.update { it.copy(createForm = form.copy(saving = true, error = null)) }
        viewModelScope.launch {
            runCatching { repo.create(form.email, form.password) }
                .onSuccess {
                    _ui.update { it.copy(createForm = null, credentials = CredentialResult(form.email.trim(), form.password, isReset = false)) }
                    load()
                }
                .onFailure { e -> _ui.update { it.copy(createForm = form.copy(saving = false, error = friendlyError(e))) } }
        }
    }

    fun toggleActive(sp: Salesperson) {
        viewModelScope.launch {
            runCatching { repo.setActive(sp.id, !sp.isActive) }
                .onSuccess { _ui.update { it.copy(message = if (sp.isActive) "Deactivated ${sp.email}" else "Activated ${sp.email}") }; load() }
                .onFailure { e -> _ui.update { it.copy(message = friendlyError(e)) } }
        }
    }

    fun openReset(sp: Salesperson) = _ui.update { it.copy(resetForm = ResetForm(sp)) }
    fun closeReset() = _ui.update { it.copy(resetForm = null) }
    fun setResetPassword(v: String) = _ui.update { it.copy(resetForm = it.resetForm?.copy(password = v, error = null)) }
    fun regenerateResetPassword() = _ui.update { it.copy(resetForm = it.resetForm?.copy(password = generatePassword(), error = null)) }

    fun confirmReset() {
        val form = _ui.value.resetForm ?: return
        if (!form.canSave) return
        _ui.update { it.copy(resetForm = form.copy(saving = true, error = null)) }
        viewModelScope.launch {
            runCatching { repo.resetPassword(form.sp.id, form.password) }
                .onSuccess {
                    _ui.update { it.copy(resetForm = null, credentials = CredentialResult(form.sp.email, form.password, isReset = true)) }
                }
                .onFailure { e -> _ui.update { it.copy(resetForm = form.copy(saving = false, error = friendlyError(e))) } }
        }
    }

    fun deleteStaff(sp: Salesperson) {
        viewModelScope.launch {
            runCatching { repo.delete(sp.id) }
                .onSuccess { _ui.update { it.copy(message = "Removed ${sp.email}") }; load() }
                .onFailure { e -> _ui.update { it.copy(message = friendlyError(e)) } }
        }
    }
}

private fun generatePassword(): String {
    val chars = "ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnpqrstuvwxyz23456789"
    val rnd = SecureRandom()
    return (1..10).map { chars[rnd.nextInt(chars.length)] }.joinToString("")
}

File: Plantbill/Plantbill/Features/More/StaffViewModel.swift
import Foundation
import Observation

private func generatePassword() -> String {
    let chars = Array("ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnpqrstuvwxyz23456789")
    return String((0..<10).map { _ in chars.randomElement()! })
}

/// Result of a create/reset action — shows the one-time credentials.
struct CredentialResult: Equatable, Identifiable {
    let email: String
    let password: String
    let isReset: Bool
    var id: String { email + password }
}

struct CreateForm: Equatable {
    var email: String = ""
    var password: String = generatePassword()
    var saving = false
    var error: String?

    var canSave: Bool { email.contains("@") && password.count >= 8 && !saving }
}

/// Reset a specific salesperson's password to a chosen (or generated) value.
struct ResetForm: Equatable {
    let sp: Salesperson
    var password: String = generatePassword()
    var saving = false
    var error: String?

    var canSave: Bool { password.count >= 8 && !saving }
}

@Observable
@MainActor
final class StaffViewModel {
    private(set) var isLoading = true
    private(set) var loadError: String?
    private(set) var staff: [Salesperson] = []

    var createForm: CreateForm?
    var resetForm: ResetForm?
    var credentials: CredentialResult?
    var message: String?

    func load() async {
        isLoading = true
        loadError = nil
        do {
            staff = try await APIClient.shared.send(Endpoint(path: "shop/users"))
            isLoading = false
        } catch let error as APIError {
            isLoading = false
            loadError = error.userMessage
        } catch {
            isLoading = false
            loadError = APIError.unknown.userMessage
        }
    }

    func openCreate() { createForm = CreateForm() }
    func closeCreate() { createForm = nil }
    func regeneratePassword() { createForm?.password = generatePassword(); createForm?.error = nil }

    func createStaff() async {
        guard var form = createForm, form.canSave else { return }
        form.saving = true
        createForm = form
        do {
            let body = try APIClient.shared.encode(SalespersonCreateRequest(email: form.email.trimmingCharacters(in: .whitespacesAndNewlines), password: form.password))
            let _: Salesperson = try await APIClient.shared.send(Endpoint(path: "shop/users", method: .post, body: body))
            credentials = CredentialResult(email: form.email.trimmingCharacters(in: .whitespacesAndNewlines), password: form.password, isReset: false)
            createForm = nil
            await load()
        } catch let error as APIError {
            form.saving = false
            form.error = error.userMessage
            createForm = form
        } catch {
            form.saving = false
            form.error = APIError.unknown.userMessage
            createForm = form
        }
    }

    func toggleActive(_ sp: Salesperson) async {
        do {
            let body = try APIClient.shared.encode(SalespersonActivateRequest(isActive: !sp.isActive))
            let _: Salesperson = try await APIClient.shared.send(Endpoint(path: "shop/users/\(sp.id)", method: .patch, body: body))
            message = sp.isActive ? "\(sp.email) deactivated." : "\(sp.email) activated."
            await load()
        } catch let error as APIError {
            message = error.userMessage
        } catch {
            message = APIError.unknown.userMessage
        }
    }

    func openReset(_ sp: Salesperson) { resetForm = ResetForm(sp: sp) }
    func closeReset() { resetForm = nil }
    func regenerateResetPassword() { resetForm?.password = generatePassword(); resetForm?.error = nil }

    func confirmReset() async {
        guard var form = resetForm, form.canSave else { return }
        form.saving = true
        resetForm = form
        do {
            let body = try APIClient.shared.encode(SalespersonResetPasswordRequest(newPassword: form.password))
            let _: Salesperson = try await APIClient.shared.send(Endpoint(path: "shop/users/\(form.sp.id)/reset-password", method: .post, body: body))
            credentials = CredentialResult(email: form.sp.email, password: form.password, isReset: true)
            resetForm = nil
        } catch let error as APIError {
            form.saving = false
            form.error = error.userMessage
            resetForm = form
        } catch {
            form.saving = false
            form.error = APIError.unknown.userMessage
            resetForm = form
        }
    }

    func deleteStaff(_ sp: Salesperson) async {
        do {
            try await APIClient.shared.sendNoContent(Endpoint(path: "shop/users/\(sp.id)", method: .delete))
            message = "\(sp.email) removed."
            await load()
        } catch let error as APIError {
            message = error.userMessage
        } catch {
            message = APIError.unknown.userMessage
        }
    }

    func dismissCredentials() { credentials = nil }
    func dismissMessage() { message = nil }
}
